File: TESTES/quicksort.c
#include <stdio.h>

void swap(int vet[], int i, int j) {
    int aux = vet[i];
    vet[i] = vet[j];
    vet[j] = aux;
}

int mediana (int vet[], int esq, int dir) {
    int meio = (esq + dir) / 2;
    if (vet[esq] > vet[meio]) swap(vet, esq, meio);
    if (vet[esq] > vet[dir]) swap(vet, esq, dir);
    if (vet[meio] > vet[dir]) swap(vet, meio, dir);
    return vet[meio];
}

void quicksort (int vet[], int esq, int dir) {
    while (esq < dir) {
        int i = esq, j = dir;
        int pivo = mediana(vet, esq, dir);
        while (i <= j) {
            while (vet[i] < pivo) i++;
            while (vet[j] > pivo) j--;
            if (i <= j) {
                swap(vet, i, j);
                i++; j--;
            }
        }
        if (j - esq < dir - i) {
            quicksort(vet, esq, j);
            esq = i;
        } else {
            quicksort(vet, i, dir);
            dir = j;
        }
    }
}

File: AEDS/OFICINA/quicksort.java
import java.util.*;

public class quicksort {
    public static Scanner sc = new Scanner (System.in);

    public static void swap(int vet[], int i, int j) {
        int aux = vet[i];
        vet[i] = vet[j];
        vet[j] = aux;
    }

    public static int partition (int[] arr, int low, int high) {
        int pivo = arr[(low + high)/2];
        int i = low, j = high;
        while (i <= j) {
            while (arr[i] < pivo) i++;
            while (arr[j] > pivo) j--;
            if (i <= j) {
                swap(arr, i, j);
                i++; j--;
            }
        }
        return i;
    }

    public static void quickSort (int[] arr, int low, int high) {
        if (low < high) {
            int pivo = partition(arr, low, high);
            quickSort(arr, low, pivo - 1);
            quickSort(arr, pivo, high);
        }
    }

    public static void main (String[] args) {
        int n = sc.nextInt();
        int[] vet = new int[n];
        for (int i = 0; i < n; i++) vet[i] = sc.nextInt();
        quickSort(vet, 0, n - 1);
        System.out.print("[ ");
        for (int i = 0; i < n; i++) System.out.print(vet[i] + " ");
        System.out.print("]");
    }
}
